Source/WebKit/UIProcess/WebProcessProxy.cpp

```diff
@@ -1255,6 +1268,32 @@ RefPtr<API::UserInitiatedAction> WebProcessProxy::userInitiatedActivity(uint64_t
     return result.iterator->value;
 }
 
+RefPtr<API::UserInitiatedAction> WebProcessProxy::userInitiatedActivity(std::optional<UUID> authorizationToken, uint64_t identifier)
+{
+    if (!UserInitiatedActionMap::isValidKey(identifier) || !identifier)
+        return nullptr;
+
+    if (authorizationToken) {
+        auto it = m_userInitiatedActionByAuthorizationTokenMap.find(*authorizationToken);
+        if (it != m_userInitiatedActionByAuthorizationTokenMap.end()) {
+            auto result = m_userInitiatedActionMap.ensure(identifier, [it = WTFMove(it)] {
+                return it->value;
+            });
+            return result.iterator->value;
+        }
+    }
+
+    return userInitiatedActivity(identifier);
+}
+
+void WebProcessProxy::consumeIfNotVerifiablyFromUIProcess(RefPtr<API::UserInitiatedAction>& action, std::optional<UUID> authToken)
+{
+    if (authToken && m_userInitiatedActionByAuthorizationTokenMap.contains(*authToken))
```

Comment:
This is a typical double hash map lookup, which we try to avoid.
```
if (authToken && m_userInitiatedActionByAuthorizationTokenMap.remove(*authToken))
    return;
action->setConsumed();
```